cmd/livenessprobe/main.go

```diff
@@ -137,12 +157,25 @@ func main() {
 	}
 
 	mux := http.NewServeMux()
-	addr := net.JoinHostPort("0.0.0.0", *healthzPort)
-	metricsManager.RegisterToServer(mux, *metricsPath)
 	metricsManager.SetDriverName(csiDriverName)
 
+	if *metricsAddress != "" && *metricsAddress != addr {
+		// Remove once --metrics-address is removed
+		metricsMux := http.NewServeMux()
+		metricsManager.RegisterToServer(metricsMux, *metricsPath)
+		go func() {
+			klog.Infof("Separate metrics ServeMux listening at %q", *metricsAddress)
+			err := http.ListenAndServe(*metricsAddress, metricsMux)
```

Comment:
So this means we start two servers. One for the metrics and the other for healthz? Is it possible to do it in one server?

I am trying to understand why `metricsManager.RegisterToServer(mux, *metricsPath)` will not work.

Reply:
And could you remind me why do we need the http-endpoint?

Reply:
> And could you remind me why do we need the http-endpoint?

This is to be consistent with all other sidecars - every sidecar will have `http-endpoint` as the flag to specify where the its HTTP server should run (e.g. https://github.com/kubernetes-csi/external-provisioner/pull/537). This also eliminates the limitation that health check and metrics must run on different ports.

> So this means we start two servers. One for the metrics and the other for healthz? Is it possible to do it in one server?

This keeps the original behavior of being able to specify different ports for health check and metrics, to maintain backward compatibility. If `metrics-address` isn't set explicitly, we do exactly as you described.

Reply:
I see, thanks! I think we might want to remove this on the next major release. But this looks good to me.

Reply:
Is it possible to do this:
If http-endpoint is set, just override the metrics-address.
If both of them is set, use http-endpoint.
If only metric-address is set, keep the current behavior

Reply:
So to make it as simple to use as possible, we can do this: combine the metrics port and healthz port together. I dont see the necessary to separate them apart. So the change can just be: if http-endpoint is set, then we just use that endpoint to expose metrics and healthz and ignore the healthz and metrics_address setting. Otherwise keep the current behavior.

Reply:
The tricky thing is currently `http-endpoint` has a default value (to inherit the behavior of `health-port`), and unfortunately the golang flag package doesn't have the ability to distinguish between something that's set vs unset.